### PR TITLE
espeak-ng: new port, version 1.50

### PR DESCRIPTION
**Pull request by danielbair**
espeak-ng: new port, version 1.50

Plus dependent ports:
pcaudiolib: new port, version 1.1
sonic: new port, version 0.2.0

#### Description

The eSpeak NG is a compact open source software text-to-speech synthesizer for Linux, Windows, Android and other operating systems. It supports more than 100 languages and accents. It is based on the eSpeak engine created by Jonathan Duddington.

This replaces the espeak port per:
https://trac.macports.org/ticket/60732

###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 10.15.5 19F101
Xcode 11.5 11E608c

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?

- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by danielbair** [comment]
RE: https://trac.macports.org/ticket/60732

**Comment by mf2k** [comment]
> RE: https://trac.macports.org/ticket/60732

Please add the reference to the Trac ticket in the commit message.

File: audio/espeak-ng/Portfile [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+                    port:rb-ronn-ng \
+                    port:rb-kramdown \
+                    port:automake \
+                    port:autoconf \
+                    port:libtool \
+                    port:pkgconfig
+
+configure.cmd       ./autogen.sh && ./configure --with-extdict-ru --with-extdict-zh --with-extdict-zhy
+build.cmd           make src/espeak-ng src/speak-ng en && make
+build.target
```
Comment (ryandesign) [inline audio/espeak-ng/Portfile:36]: Overriding build.cmd in this way is unusual. Why is this necessary? Does `build.target-append src/espeak-ng src/speak-ng en docs` not work?

`-j8` should not be hardcoded here. If you absolutely must override build.cmd to contain multiple make command invocations, you can get the correct parallel build value using `-j${build.jobs}`.

File: audio/espeak-ng/Portfile [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+build.args-append   -i -k prefix=${prefix}
+destroot.args       {*}${build.args}
+
+test.run            yes
+test.cmd            ESPEAK_DATA_PATH=`pwd` LD_LIBRARY_PATH=src src/espeak-ng "Testing." -w test.wav
```
Comment (ryandesign) [inline audio/espeak-ng/Portfile:41]: Why set LD_LIBRARY_PATH? macOS doesn't use that variable; it uses DYLD_LIBRARY_PATH.

Normally specifying environment variables to the test phase should be done with test.env. However I had [trouble getting that to work for espeak](https://trac.macports.org/ticket/60722). If the same is happening here, then setting them in test.cmd is fine.

File: audio/pcaudiolib/Portfile [hunk 3]
```diff
@@ -0,0 +1,36 @@
+# -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:ft=tcl:et:sw=4:ts=4:sts=4
+
+PortSystem          1.0
+PortGroup           github 1.0
+
+name                pcaudiolib
+github.setup        espeak-ng pcaudiolib 1.1
+version             1.1
+revision            1
```
Comment (ryandesign) [inline audio/pcaudiolib/Portfile:9]: Don't set name or version; github.setup sets them for you. Set revision to 0.

File: audio/pcaudiolib/Portfile [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+platforms           darwin
+maintainers         nomaintainer
+license             GPL-3+
+
+description         Portable C Audio Library
+
+long_description    Portable C Audio Library
+
+homepage            https://github.com/espeak-ng/pcaudiolib
+distname            ${version}
```
Comment (ryandesign) [inline audio/pcaudiolib/Portfile:20]: Don't set homepage or distname; github.setup sets them for you.

File: audio/pcaudiolib/Portfile [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+                    port:libtool \
+                    port:pkgconfig
+
+configure.cmd        ./autogen.sh && ./configure
+build.args-append    prefix=${prefix}
+destroot.args        {*}${build.args}
+
+test.run            yes
+test.cmd            make check
+test.target
```
Comment (ryandesign) [inline audio/pcaudiolib/Portfile:36]: Don't put the test target into test.cmd; put it into test.target. The default for test.cmd is make so you can remove the test.cmd line and just set `test.target check`.

File: audio/sonic/Portfile [hunk 6]
```diff
@@ -0,0 +1,46 @@
+# -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:ft=tcl:et:sw=4:ts=4:sts=4
+
+PortSystem          1.0
+PortGroup           github 1.0
+PortGroup           makefile 1.0
+
+name                sonic
+github.setup        espeak-ng sonic 0.2.0 release-
+version             0.2.0
+revision            1
```
Comment (ryandesign) [inline audio/sonic/Portfile:10]: Don't set name or version; github.setup sets them for you. Set revision to 0.

File: audio/sonic/Portfile [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+platforms           darwin
+maintainers         nomaintainer
+license             GPL-3+
+
+description         Simple library to speed up or slow down speech
+
+long_description    Simple library to speed up or slow down speech
+
+homepage            https://github.com/espeak-ng/sonic
+distname            release-${version}
```
Comment (ryandesign) [inline audio/sonic/Portfile:21]: Don't set homepage or distname; github.setup sets them for you.

File: audio/sonic/Portfile [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+configure.cxxflags-append -std=c++98
+
+makefile.override   PREFIX
+
+build.args-append   USE_SPECTROGRAM=0
+destroot.args       {*}${build.args}
+
+test.run            yes
+test.cmd            make check
+test.target
```
Comment (ryandesign) [inline audio/sonic/Portfile:46]: Don't set test.cmd; Set `test.target check`.

File: audio/pcaudiolib/Portfile [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+description         Portable C Audio Library
+
+long_description    Portable C Audio Library
+
+homepage            https://github.com/espeak-ng/pcaudiolib
+distname            ${version}
+checksums           rmd160  f68876d3346734e993b9d5ebfc3ef48de9161851 \
+                    sha256  449d69c3fd4ed0136d357fd42a73accb276da9b35a19258a379e3ac7ac88c148 \
+                    size    34082
```
Comment (ryandesign) [inline audio/pcaudiolib/Portfile:23]: Please use `github.tarball_from archive`. This will require changing the checksums as well.

File: audio/sonic/Portfile [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+PortGroup           makefile 1.0
+
+name                sonic
+github.setup        espeak-ng sonic 0.2.0 release-
+version             0.2.0
+revision            1
+categories          audio
+platforms           darwin
+maintainers         nomaintainer
+license             GPL-3+
```
Comment (ryandesign) [inline audio/sonic/Portfile:14]: sonic is under the Apache-2 license, not GPL-3+.

File: audio/sonic/Portfile [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+description         Simple library to speed up or slow down speech
+
+long_description    Simple library to speed up or slow down speech
+
+homepage            https://github.com/espeak-ng/sonic
+distname            release-${version}
+checksums           rmd160  fd728ede06bf9f2b5ff8257455c506341e5ee620 \
+                    sha256  66ae97ccd39450734a3819c342683fdb1698eb2322c011b95fcf7e1959caa1db \
+                    size    5340290
```
Comment (ryandesign) [inline audio/sonic/Portfile:24]: Please use `github.tarball_from archive`. This will require changing the checksums as well.

File: audio/pcaudiolib/Portfile [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+depends_lib         port:automake \
+                    port:autoconf \
+                    port:libtool \
+                    port:pkgconfig
+
+configure.cmd        ./autogen.sh && ./configure
```
Comment (ryandesign) [inline audio/pcaudiolib/Portfile:30]: `configure.cmd       ./autogen.sh && ./configure` is acceptable although it is preferable to use:
```
use_autoreconf yes
autoreconf.cmd ./autogen.sh
```

If you make this change, make sure you append the autoconf, automake and libtool build dependencies *after* this line. If you have them before this line, they will be removed.

File: audio/espeak-ng/Portfile [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+name                espeak-ng
+github.setup        espeak-ng espeak-ng 1.50
+version             1.50
+revision            1
+categories          audio
+platforms           darwin
+maintainers         nomaintainer
+license             GPL-3+
+
+description         eSpeak NG is a multi-lingual software speech synthesizer
```
Comment (ryandesign) [inline audio/espeak-ng/Portfile:15]: Normally the description should be a sentence fragment e.g. `multi-lingual software speech synthesizer`

File: audio/pcaudiolib/Portfile [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+version             1.1
+revision            1
+categories          audio
+platforms           darwin
+maintainers         nomaintainer
+license             GPL-3+
+
+description         Portable C Audio Library
+
+long_description    Portable C Audio Library
```
Comment (ryandesign) [inline audio/pcaudiolib/Portfile:17]: The long description should be full sentences, such as `${name} is a portable C audio library.` That's a fairly short long description though. If more information is available, you could add a sentence or two more.

File: audio/espeak-ng/Portfile [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+                    port:rb-kramdown \
+                    port:automake \
+                    port:autoconf \
+                    port:libtool \
+                    port:pkgconfig
+
+configure.cmd       ./autogen.sh && ./configure --with-extdict-ru --with-extdict-zh --with-extdict-zhy
+build.cmd           make src/espeak-ng src/speak-ng en && make
+build.target
+build.args-append   -i -k prefix=${prefix}
```
Comment (ryandesign) [inline audio/espeak-ng/Portfile:37]: Looks like `-i` ignores errors and `-k` keeps going. Why is this necessary? I would not think that we would want to ignore errors.

Is specifying prefix really necessary here? The build system should be remembering the prefix that was specified at configure time. If the build system is broken and is failing to do that, please make sure there's an upstream bug report about that and include the URL in a comment in the portfile.

File: audio/espeak-ng/Portfile [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+depends_lib         port:pcaudiolib \
+                    port:sonic
+depends_build       port:rb26-ronn-ng \
+                    port:rb-kramdown \
+                    port:automake \
+                    port:autoconf \
+                    port:libtool \
+                    port:pkgconfig
+
+configure.cmd       ./autogen.sh && RONN=${prefix}/bin/ronn-2.6 ./configure --with-extdict-ru --with-extdict-zh --with-extdict-zhy
```
Comment (ryandesign) [inline audio/espeak-ng/Portfile:34]: Args (`--with-extdict-ru` ...) should not be specified in configure.cmd. Specify them in `configure.args` instead.

Environment variables (`RONN=...`) should not be specified in configure.cmd. Specify them in `configure.env` instead.

*If you need to autoreconf* then `configure.cmd       ./autogen.sh && ./configure` is acceptable although it is preferable to use:
```
use_autoreconf yes
autoreconf.cmd ./autogen.sh
```

If you make this change, make sure you append the autoconf, automake and libtool build dependencies *after* this line. If you have them before this line, they will be removed.

However, *do you really need to autoreconf*? If you were only doing this because the repository and therefore the auto-generated tarball does not contain a configure script, then switching to the release tarball as suggested elsewhere solves that problem.

File: audio/espeak-ng/Portfile [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+homepage            https://github.com/espeak-ng/espeak-ng
+distname            ${version}
+checksums           rmd160  bd3dd4e17863bb1aa3a964517d888ed9997f0e3e \
+                    sha256  f0505366d5391f1291a9c5cc0816461719f46eaed84cfbf57fb81b0dc381de21 \
+                    size    13665736
+
+depends_lib         port:pcaudiolib \
+                    port:sonic
+depends_build       port:rb26-ronn-ng \
+                    port:rb-kramdown \
```
Comment (ryandesign) [inline audio/espeak-ng/Portfile:28]: Is it necessary to involve two different old ruby versions in the build process? rb-kramdown uses ruby (1.8.6) and rb26-ronn-ng uses ruby26 (2.6.6). ruby27 (2.7.1) is the latest version and should be used if possible. I realize neither kramdown nor ronn-ng are available in rb27 flavors in MacPorts right now so this may be a to-do item for later.

File: audio/pcaudiolib/Portfile [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+                    size    34082
+
+depends_build       port:automake \
+                    port:autoconf \
+                    port:libtool \
+                    port:pkgconfig
+
+configure.cmd        ./autogen.sh && ./configure
+build.args-append    prefix=${prefix}
+destroot.args        {*}${build.args}
```
Comment (ryandesign) [inline audio/pcaudiolib/Portfile:32]: Is specifying prefix really necessary here? The build system should be remembering the prefix that was specified at configure time. If the build system is broken and is failing to do that, please make sure there's an upstream bug report about that and include the URL in a comment in the portfile.

File: audio/sonic/files/patch-sonic-mac.diff [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
++
++# Set this to empty if not using spectrograms.
+ 
+-all: sonic libsonic.so.$(LIB_TAG) libsonic.a
++all: sonic $(LIB_NAME)$(LIB_TAG) libsonic.a
+ 
+-sonic: wave.o main.o libsonic.so.$(LIB_TAG)
+-	$(CC) $(CFLAGS) -o sonic wave.o main.o libsonic.so.$(LIB_TAG)
++sonic: wave.o main.o libsonic.a
++	$(CC) $(CFLAGS) -o sonic wave.o main.o libsonic.a -lm $(FFTLIB)
```
Comment (ryandesign) [inline audio/sonic/files/patch-sonic-mac.diff:53]: You've changed sonic from using the shared library to using the static library. This means a copy of the static library will be embedded in the executable, which wastes disk space and is generally undesirable. This should continue to use the shared (dynamic on macOS) library.

This will require the LD_LIBRARY_PATH (DYLD_LIBRARY_PATH on macOS) environment variable to be set properly in the check target.

File: audio/sonic/files/patch-sonic-mac.diff [hunk 20]
```diff
@@ -0,0 +1,119 @@
+--- Makefile.orig	2020-06-30 13:32:50.000000000 +0700
++++ Makefile	2020-06-30 14:09:51.000000000 +0700
```
Comment (ryandesign) [inline audio/sonic/files/patch-sonic-mac.diff:2]: I hope you're sharing all of these changes with the developers of sonic. If so, or if you got these changes from a patch developed by someone else, please mention at least the URL at the top of the patchfile.

File: audio/espeak-ng/Portfile (hunk 17 above)
Comment (danielbair) [reply audio/espeak-ng/Portfile:28]: rb-ronn-ng only contains one file (README)
rb19-ronn uses ruby 1.9
rb-kramdown uses ruby 1.8 but throws this error:
/opt/local/lib/ruby/vendor_ruby/1.8/kramdown/document.rb:40:in `data_dir': kramdown data directory not found! This is a bug, please report it! (RuntimeError)

...thus the build.args -i -k

File: audio/espeak-ng/Portfile (hunk 15 above)
Comment (danielbair) [reply audio/espeak-ng/Portfile:37]: I found a way to avoid -i and -k args.

File: audio/espeak-ng/Portfile (hunk 17 above)
Comment (danielbair) [reply audio/espeak-ng/Portfile:28]: I found a way to avoid the -i and -k args

File: audio/sonic/files/patch-sonic-mac.diff (hunk 19 above)
Comment (danielbair) [reply audio/sonic/files/patch-sonic-mac.diff:53]: That was from git head from the developer. I will check if the dylib can still be used here.

File: audio/sonic/files/patch-sonic-mac.diff (hunk 20 above)
Comment (danielbair) [reply audio/sonic/files/patch-sonic-mac.diff:2]: I will share back to the developer. The day I was working on this github was having problems (per their status page) and I couldn't file any issues or pull requests.

File: audio/sonic/files/patch-sonic-mac.diff (hunk 19 above)
Comment (danielbair) [reply audio/sonic/files/patch-sonic-mac.diff:53]: Ok, it builds and tests with the dylib here.

File: audio/pcaudiolib/Portfile (hunk 14 above)
Comment (danielbair) [reply audio/pcaudiolib/Portfile:17]: Updated from the README, but it is still somewhat short.

File: audio/pcaudiolib/Portfile (hunk 12 above)
Comment (danielbair) [reply audio/pcaudiolib/Portfile:30]: Got it, Thanks!

File: audio/sonic/Portfile (hunk 10 above)
Comment (danielbair) [reply audio/sonic/Portfile:14]: Oops. Fixed!

File: audio/espeak-ng/Portfile [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+
+description         eSpeak NG is a multi-lingual software speech synthesizer
+
+long_description    eSpeak NG is a multi-lingual software speech synthesizer
+
+homepage            https://github.com/espeak-ng/espeak-ng
+distname            ${version}
+checksums           rmd160  bd3dd4e17863bb1aa3a964517d888ed9997f0e3e \
+                    sha256  f0505366d5391f1291a9c5cc0816461719f46eaed84cfbf57fb81b0dc381de21 \
+                    size    13665736
```
[the remark this answers is not in the document]
Comment (danielbair) [reply audio/espeak-ng/Portfile:23]: I could not use the release tarball as neither autogen.sh nor configure would work from it.

File: audio/espeak-ng/Portfile [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+platforms           darwin
+maintainers         nomaintainer
+license             GPL-3+
+
+description         eSpeak NG is a multi-lingual software speech synthesizer
+
+long_description    eSpeak NG is a multi-lingual software speech synthesizer
+
+homepage            https://github.com/espeak-ng/espeak-ng
+distname            ${version}
```
[the remark this answers is not in the document]
Comment (danielbair) [reply audio/espeak-ng/Portfile:20]: Got it. Thanks!

File: audio/espeak-ng/Portfile [hunk 23]
```diff
@@ -0,0 +1,42 @@
+# -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:ft=tcl:et:sw=4:ts=4:sts=4
+
+PortSystem          1.0
+PortGroup           github 1.0
+
+name                espeak-ng
+github.setup        espeak-ng espeak-ng 1.50
+version             1.50
+revision            1
```
[the remark this answers is not in the document]
Comment (danielbair) [reply audio/espeak-ng/Portfile:9]: Got it. Thanks!

**Comment by danielbair** [comment]
xcode9.4 fails with this:
Error: Portfile is from the future - check date and time of your system
Error: See /opt/local/var/macports/logs/_Users_travis_build_macports_macports-ports_math_fftw-3/fftw-3/main.log for details.
Error: Follow https://guide.macports.org/#project.tickets to report a bug.
Error: Processing of port fftw-3 failed
Fetching archive for dependency 'fftw-3' with variants '' failed, aborting.

File: audio/espeak-ng/Portfile (hunk 16 above)
Comment (danielbair) [reply audio/espeak-ng/Portfile]: I could not get the configure from the release tarball to work...

**Comment by ryandesign** [comment]
> xcode9.4 fails with this:
> Error: Portfile is from the future - check date and time of your system

https://travis-ci.community/t/osx-system-clock-occasionally-goes-back-by-20-minutes/6592/5

**Comment by danielbair** [comment]
@ryandesign I have a new port I am waiting to submit that depends on this espeak-ng port. Is there something else needed before merging this?

**Comment by kencu** [comment]
I can't figure out if I can merge this.

File: audio/espeak-ng/Portfile (hunk 15 above)
Comment (danielbair) [reply audio/espeak-ng/Portfile]: I finally found out that this was related to espeak-ng not supporting parallel building. My solution on my system was to set '-j1', but is there another more preferable way to disable parallel building?

Comment (mf2k) [reply audio/espeak-ng/Portfile]: `use_parallel_build no
`

[Here is the reference]([url](https://guide.macports.org/#reference.phases.build)) from the guide.
